src/algebra/squarefree.lean

```diff
@@ -204,4 +204,55 @@ lemma sum_divisors_filter_squarefree {n : ℕ} (h0 : n ≠ 0)
 by rw [finset.sum_eq_multiset_sum, divisors_filter_squarefree h0, multiset.map_map,
     finset.sum_eq_multiset_sum]
 
+lemma sq_mul_squarefree {n : ℕ} (hn : 0 < n) :
+  ∃ a b : ℕ, 0 < a ∧ 0 < b ∧ b ^ 2 * a = n ∧ squarefree a :=
+begin
+  let S := {s ∈ finset.range (n + 1) | s ∣ n ∧ ∃ x, s = x ^ 2},
+  have hSne : S.nonempty,
+  { use 1,
+    have h1 : 0 < n ∧ ∃ (x : ℕ), 1 = x ^ 2 := ⟨hn, ⟨1, (one_pow 2).symm⟩⟩,
+    simpa [S] },
+
+  let s := finset.max' S hSne,
+  have hs : s ∈ S := finset.max'_mem S hSne,
+  simp only [finset.sep_def, S, finset.mem_filter, finset.mem_range] at hs,
+  obtain ⟨hsn1, ⟨a, hsa⟩, ⟨b, hsb⟩⟩ := hs,
+
+  rw hsa at hn,
+  obtain ⟨hlts, hlta⟩ := canonically_ordered_semiring.mul_pos.mp hn,
+  rw hsb at hsa hn hlts,
+  have hltb : 0 < b := by rwa [pow_two, canonically_ordered_semiring.mul_pos, and_self] at hlts,
+
+  refine ⟨a, b, hlta, hltb, hsa.symm, _⟩,
+  rintro x ⟨y, hy⟩,
+  by_contradiction,
+  rw nat.is_unit_iff at h,
+
+  have hx₀ : ¬(x = 0) := λ hyp, by simp * at *,
+  have hx₂ : x > 1 := nat.one_lt_iff_ne_zero_and_ne_one.mpr ⟨hx₀, h⟩,
+
+  set z := (b * x) ^ 2 with hz,
+  have hz₀ : z > s,
+  { rw [hz, hsb, mul_pow b x 2, gt_iff_lt],
+    refine lt_mul_of_one_lt_right hlts _,
+    rw pow_two, exact one_lt_mul' hx₂ hx₂ },
+  have hz₁ : z ∣ n, { use y, rw [hsa, hy, hz], ring },
+  have hz₂ : z ∈ S,
+  { simp only [finset.sep_def, S, finset.mem_filter, finset.mem_range],
+    exact ⟨lt_succ_iff.mpr (le_of_dvd (by rwa ← hsa at hn) hz₁), hz₁, ⟨b * x, rfl⟩⟩ },
+  have hz₃ : z ≤ s := finset.le_max' S z hz₂,
+
+  exact lt_le_antisymm hz₀ hz₃,
```

Comment:
May I suggest revamping this part of the proof as follows? It is more direct imo.
```suggestion
  rw nat.is_unit_iff,
  by_contra hx,
  refine lt_le_antisymm _ (finset.le_max' S ((b * x) ^ 2) _),
  { simp_rw [S, finset.sep_def, finset.mem_filter, finset.mem_range],
    refine ⟨lt_succ_iff.mpr (le_of_dvd (by rwa ← hsa at hn) _), _, ⟨b * x, rfl⟩⟩;
    use y; rw [hsa, hy]; ring },
  { convert lt_mul_of_one_lt_right hlts
      (one_lt_pow 2 x zero_lt_two (one_lt_iff_ne_zero_and_ne_one.mpr ⟨λ h, by simp * at *, hx⟩)),
    rw mul_pow },
```